### DISCUSSION TITLE
Merging multiple Store slices with Typescript

### DISCUSSION THREAD

**Discussion by P1llus** [pmndrs/zustand#2371] [Q&A] [answered]
Hey Zustand community!

TLDR: Is there an easy or clean way to merge multiple sliced stores without having to import way to many types etc? More description at the bottom.

I have been using zustand for plenty of smaller JS projects for a longer time, always so easy to use and I have a simple "skeleton" I reuse for all my quick mockup projects etc.

Starting a new project now, this time with Typescript (TSX) and react rather than just JS/JSX, using the latest stable version from zustand.

I started by going to the Typescript section, I see there is now a new major version and a revised typescript guide: https://docs.pmnd.rs/zustand/guides/typescript

As slices is something I have always used in my projects, I also checked out that section: https://docs.pmnd.rs/zustand/guides/typescript#slices-pattern
and of course the advanced section:
https://docs.pmnd.rs/zustand/guides/slices-pattern

This allowed me to see what changes I would have to make so that I can make a similar pattern I have used on my JS projects, here is an example from my GlobalState in JS projects, where I have many sliced stores in different files, imported and merged into the global state:

```javascript
import { create } from 'zustand';
import { formState, logSamplesState, tokenState, gptState } from './InputFormState';
import { ecsTablePopover, ecsTableFields } from './EcsTableState';
import {
  pipelineErrors,
  pipelineState,
  pipelineRunResults,
  pipelineStats,
  pipelineSteps,
  pipelineSkippedSteps,
  pipelineStatsTotal,
  pipelineCopy,
  pipelineBadgePopover,
} from './IngestPipelineState';

export const useGlobalState = create()((...a) => ({
  ...formState(...a),
  ...logSamplesState(...a),
  ...tokenState(...a),
  ...pipelineErrors(...a),
  ...pipelineRunResults(...a),
  ...pipelineStats(...a),
  ...pipelineSteps(...a),
  ...pipelineSkippedSteps(...a),
  ...pipelineStatsTotal(...a),
  ...pipelineState(...a),
  ...pipelineCopy(...a),
  ...pipelineBadgePopover(...a),
  ...gptState(...a),
  ...ecsTablePopover(...a),
  ...ecsTableFields(...a),
}));
```

Now when I started creating a similar one in typescript based on the docs, it started out with something liket his:
Define a few stores:

```typescript
interface EcsMappingFormState {
  packageName: string;
  dataStreamName: string;
  setEcsMappingFormState: (key: string, value: string) => void;
}

interface EcsMappingSamplesState {
  samples: string[];
  increaseSample: () => void;
  decreaseSample: () => void;
  setSample: (index: number, sample: string) => void;
  addSample: (sample: string) => void;
}

const ecsMappingFormState: StateCreator<EcsMappingFormState & EcsMappingSamplesState, [], [], EcsMappingFormState> = (
  set: any,
): EcsMappingFormState => ({
  packageName: '',
  dataStreamName: '',
  setEcsMappingFormState: (key, value) => set(() => ({ [key]: value })),
});
```

The question I have is this, when I want to merge all my sliced stores into a single one, do I then have to import the types from every single store? That will be a large list of imported interfaces and stores, and the `create` function at the end would also need all of their types passed in as its first argument as well? That will be a very long list of interfaces/types?

Example would be:
```
export const useGlobalState = create<EcsMappingFormState & EcsMappingSamplesState>()((...a) => ({
  ...ecsMappingFormState(...a),
  ...ecsMappingSamplesState(...a),
}));
```

But what happens when I have 15-20 slices? Do I need to pass all the 15-20 interfaces in with `create<>`?

**Comment by dbritto-dev** [comment, marked as answer]
@P1llus yes, you do
